### ISSUE TITLE
Errors occurred during the testing phase

### ISSUE THREAD

**Issue by crrasjtu** [LMD0311/HERMES#8]
Thank you for sharing your code and pretrained weight.

I encountered a problem while running testing with the hermes_final.pth and hermes_stage.py in this repository. The program crashes with the following error:

<img width="1199" height="306" alt="Image" src="https://github.com/user-attachments/assets/2fcef333-ae00-4095-aae6-bcbae4725dc7" />

After inspecting the core dump using gdb, I found that the error occurs in the CUDA cuBLAS library function `cublasLtTSTMatmulAlgoGetHeuristic()`. This happens during the testing phase, when using both the official code and the official weights, without any modifications. 

<img width="1202" height="444" alt="Image" src="https://github.com/user-attachments/assets/2c533078-1b22-4ae2-9bf6-8bf81db1e48a" />

I have further investigated the issue, and found that the crash during evaluation happens inside the LLM's decoder layer. 

![Image](https://github.com/user-attachments/assets/c08d1998-6bbb-48bd-9ede-91c47273aafd)

Additionally, I would like to mention that I was able to successfully run training for stage1, stage2-1, and stage2-2 without any issues. However, when training stage3, I encountered NaN values starting from the second epoch. 

<img width="1379" height="459" alt="Image" src="https://github.com/user-attachments/assets/3dbbb39e-cb48-4c98-9584-403e4c84257a" />

Could you please help me identify what might be causing this issue? Is there any known compatibility problem with specific hardware or software versions? Any suggestions for further debugging would also be appreciated.

My environment:
- Python: 3.10
- PyTorch: 2.1.0+cu121
- CUDA: 12.3
- cuDNN: 9.0.0
- GPU: H20

Thank you!

**Comment by LMD0311** [comment]
Thanks for your interest in our work!

I apologize that I haven't encountered this specific issue previously, though I did experience a `no kernel image` problem on H20 in the past. I found that CUDA 11.8 was the most stable version. Additionally, including `TORCH_CUDA_ARCH_LIST='9.0' TORCH_SHOW_CPP_STACKTRACES=1 CUDA_ENABLE_COREDUMP_ON_EXCEPTION=1 TORCH_USE_CUDA_DSA=1 CUDA_LAUNCH_BLOCKING=1` during compilation seemed to be helpful. I regret that I cannot offer further immediate solutions.

Regarding the NaN issue, try reducing the learning rate for stage 3. We used 32 GPUs for training, and since your training time suggests you're using 8, this adjustment might stabilize the process.

**Comment by crrasjtu** [comment]
Thank you for your helpful suggestions!

We found that after switching to 32 GPUs, the NaN issue was resolved.
As for the test problem, after further investigation, we discovered that it was caused by the use of bfloat16 in the linear layer of llm. Upgrading PyTorch to version 2.7 allowed us to successfully correct this error.

<img width="1190" height="488" alt="Image" src="https://github.com/user-attachments/assets/d37540af-c425-432c-aff7-2b717ee61182" />

**Comment by LMD0311** [comment]
Thanks for your feedback! I am closing this issue. Please feel free to reopen it!
